docs/guides/extending-beluga.md

```diff
@@ -1,3 +1,402 @@
+- `Eigen` and `Sophus` for linear algebra and pose handling.
+
+```cpp
+namespace beluga {
+
+// Example of a simple range-based sensor model.
+class RangeSensorModel {
+ public:
+  using state_type = Sophus::SE2d;        // Represents the particle's state as SE(2) pose.
+  using weight_type = double;             // Particle weight, based on measurement likelihood.
+  using measurement_type = double;        // Represents a distance measurement.
+```
+
+Here, the `RangeSensorModel` class is defined. `state_type` is specified as `Sophus::SE2d`, representing the robot’s pose; `weight_type` is a `double`, indicating the computed weight for each particle; and `measurement_type` is `double`, representing the sensor measurement.
+
+```cpp
+  // Constructor to initialize sensor noise parameters.
+  RangeSensorModel(double sensor_stddev)
+      : sensor_noise_{0.0, sensor_stddev} {}
+```
+The constructor initializes a normal distribution, `sensor_noise_`, to simulate measurement noise. This distribution is centered at `0.0` with a standard deviation defined by `sensor_stddev`.
+
+```cpp
+  // Callable function to produce a StateWeightingFunction based on a measurement.
+  auto operator()(measurement_type measurement) const {
+    // Returns a lambda satisfying the StateWeightingFunction requirements.
+    return [this, measurement](const state_type& current_state) -> weight_type {
+      // Calculate the expected sensor position based on the robot's pose.
+      double expected_measurement = current_state.translation().norm();
+```
+The `operator()` function returns a lambda. This lambda calculates the weight of a particle based on the error between the expected and actual measurements.
+
+```cpp
+      // Compute the distance between the expected and actual measurement.
+      double distance = (expected_measurement - measurement).norm();
+
+      // Calculate weight based on Gaussian likelihood.
+      double weight = std::exp(-0.5 * std::pow(distance / sensor_noise_.stddev(), 2));
+
+      return weight;
+    };
+  }
+```
+A Gaussian kernel is used to compute the weight, assigning higher values to states closer to the measurement.
+
+### Next steps
+
+For a real-world implementation, see [`beluga::LikelihoodFieldModel`](https://ekumen-os.github.io/beluga/packages/beluga/docs/_doxygen/generated/reference/html/classbeluga_1_1LikelihoodFieldModel.html)'s' implementation.
+
+## Implementing Estimation Algorithms
+
+Estimation algorithms in Beluga are responsible for synthesizing the most likely estimate of a robot's state from a set of weighted particles. Statistical location and dispersion measures such as sampled means and variances are typical.
+
+Estimation algorithms are but one application of the [_niebloid_ concept](https://en.cppreference.com/w/Template:cpp/ranges/niebloid) in Beluga. While this section focuses on estimation algorithms, it is worth noting that niebloids are used more broadly across the library as a sane approach to customization.
```

Comment:
```suggestion
Estimation algorithms are implemented as functions that take a particle range as input. In this library, instead of implementing them as free functions, we implement them as [_niebloids_](https://en.cppreference.com/w/Template:cpp/ranges/niebloid) (global instances of function objects that disable ADL). While this section focuses on estimation algorithms, it is worth noting that niebloids are used more broadly across the library as a sane approach to customization.
```